src/merkle/tiered_smt/storage.rs

```diff
@@ -0,0 +1,150 @@
+use super::{BTreeMap, CanonicalWord, NodeIndex, NodeType, Vec, Word};
+use core::convert::Infallible;
+
+/// In-memory storage is infallible.
+pub type StorageError = Infallible;
+
+// TIERED SPARSE MERKLE TREE STORAGE
+// ================================================================================================
+
+/// A storage implementation for the tiered sparse merkle tree.
+#[derive(Debug, Default)]
+pub struct Storage {
+    types: BTreeMap<NodeIndex, NodeType>,
+    nodes: BTreeMap<NodeIndex, Word>,
+    keys: BTreeMap<CanonicalWord, NodeIndex>,
+    upper_leaf_keys: BTreeMap<NodeIndex, CanonicalWord>,
+    leaf_values: BTreeMap<CanonicalWord, Word>,
+    ordered_leaves: BTreeMap<u64, Vec<CanonicalWord>>,
+}
```

Comment:
The proposed `bottom_leaves` is a solid improvement, but as for the others, it's not only a matter of performance but also simplicity.

If we, for example, want to know what is a node type, currently, we just `self.storage.get_type`. If we remove that mapping, then we need to `if self.nodes.contains_key -> Node else if self.upper_leaves.contains_key || self.bottom_leaves.contains_key -> Leaf else Empty`. This complexity will extend to the other operations as well, when it could be a simple `get`

I propose we merge the two ideas, picking the best of each. wdyt?

```rust
#[derive(Debug, Default)]
pub struct Storage {
    types: BTreeMap<NodeIndex, NodeType>,
    nodes: BTreeMap<NodeIndex, Word>,
    keys: BTreeMap<CanonicalWord, NodeIndex>,
    upper_leaf_keys: BTreeMap<NodeIndex, CanonicalWord>,
    upper_leaf_values: BTreeMap<CanonicalWord, Word>,
    bottom_leaves: BTreeMap<u64, Vec<Leaf>>,
}
```

Reply:
My thinking was that `get_type()` method would be the only one impacted as we can put all the logic for determining the type there, and then everywhere else use this method. Or would this not work?

In my mind this simplifies things as there are fewer things we need to update during insertions/deletions, which were I expect the bulk of complexity is.

Reply:
I'm not sure we are achieving simplification that way. Another example is a query to the storage on where a given key is. Currently, we `storage.keys.get`; if we remove that mapping, we will need to iterate all key/value pairs of the upper leaves, fallback to the bottom leaves and iterate their vectors, until we find a matching key.